### PR TITLE
Enchanced obstacles generation and collistion

### LINKED ISSUE #3: Store information about the generated obstacles
We need to store some information about the obstacles (their positions) so when we move the player, we can check if a collision will appear.

image

### LINKED ISSUE #12: Block the collision between main player and obstacle
Under no circumstances the main player should ever step over an obstacle.

This issue depends on #3.

### PR DESCRIPTION
**Pull request by ChristopherPetrov**
Instead of using random positions, the player now can generate obstacles right to himself. The edges are randomized.

Screenshot 2023-04-27 121154

### CODE REVIEW

**Review by TonyTroeff — COMMENTED** [review]
Try with rebuild.
If this did not help, run `dotnet restore` or `dotnet restore --force` at the root folder.

File: Labyrinth.Console/Constants.cs [hunk 1]
```diff
@@ -3,5 +3,7 @@
     internal static class Constants
     {
         public const char PlayerSymbol = '*';
+
+        public const int systemRows = 1;
```
Comment (TonyTroeff) [inline Labyrinth.Console/Constants.cs]: Use PascalCase

File: Labyrinth.Console/MapGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading.Tasks;
+using System.Xml;
+
+namespace Labyrinth.Console
+{
+    public class MapGenerator
+    {
+        private static int leftBorder => 33;
```
Comment (TonyTroeff) [inline Labyrinth.Console/MapGenerator.cs]: Too much magic over here.

File: Labyrinth.Console/MapGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        private static int rightBorder => System.Console.LargestWindowWidth - 20;
```
Comment (TonyTroeff) [inline Labyrinth.Console/MapGenerator.cs]: We should not depend on the Console.

File: Labyrinth.Console/MapGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        private static int topBorder => Constants.systemRows - 1;
+        private static int botBorder => System.Console.LargestWindowHeight - 6;
+
+        public MapGenerator(Dictionary<Coordinates, Obstacle> obstaclesDict, Dictionary<ObstacleEdges, char> edgeSymbolsMap)
+        {
+            this._obstaclesDict = obstaclesDict;
+            this._edgeSymbolsMap = edgeSymbolsMap;
+        }
+
+        Dictionary<Coordinates, Obstacle> _obstaclesDict { get; }
```
Comment (TonyTroeff) [inline Labyrinth.Console/MapGenerator.cs]: Either make this a field, or use the appropriate naming convention (PascalCase)

File: Labyrinth.Console/MapGenerator.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        Dictionary<ObstacleEdges, char> _edgeSymbolsMap { get; }
+
+        public void RenderObstacle(Obstacle obstacle)
```
Comment (TonyTroeff) [inline Labyrinth.Console/MapGenerator.cs]: It is good that we have extracted such a method. However, this is not the correct place for it. I would consider moving it to a static class.

File: Labyrinth.Console/MapGenerator.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        }
+
+        public void GenerateMapBorders()
+        {
+            for (int x = leftBorder + 1; x < rightBorder; x++)
+            {
+                for (int y = topBorder; y < botBorder; y++)
+                {
+                    int whatEdges = 0;
+                    if (x + 1 == rightBorder) whatEdges += 4;
```
Comment (TonyTroeff) [inline Labyrinth.Console/MapGenerator.cs]: The magic continues.

File: Labyrinth.Console/MapGenerator.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+                Coordinates currentObstacleCoordinates = new Coordinates { X = randomObstacleX, Y = randomObstacleY };
+                Obstacle currentObstacle = new Obstacle(currentObstacleCoordinates, randomObstacleEdges);
+
+                RenderObstacle(currentObstacle);
+            }
+        }
+    }
+}
+//Top = 1, ═
```
Comment (TonyTroeff) [inline Labyrinth.Console/MapGenerator.cs]: No need for these comments.

File: Labyrinth.Console/MapGenerator.cs (hunk 6 above)
Comment (TonyTroeff) [reply Labyrinth.Console/MapGenerator.cs]: Use bitwise OR wherever this is possible.

Comment (TonyTroeff) [reply Labyrinth.Console/MapGenerator.cs]: Potentially, improve the complexity of this algorithm.

**Review by TonyTroeff — CHANGES_REQUESTED** [review]

File: Labyrinth.Console/Controllers/IFlowController.cs [hunk 8]
```diff
@@ -0,0 +1,12 @@
+﻿using Labyrinth.Console.Obstacles;
+
+namespace Labyrinth.Console.Controllers
+{
+    public interface IFlowController
+    {
+        void RenderObstacle(Obstacle obstacle);
+
+        public void QuickObstacleGeneration(Coordinates currentCoordinates, ObstacleEdges currentEdges);
```
Comment (TonyTroeff) [inline Labyrinth.Console/Controllers/IFlowController.cs]: I think this method is not necessary.

File: Labyrinth.Console/MapGenerator.cs [hunk 9]
```diff
@@ -0,0 +1,140 @@
+﻿using Labyrinth.Console.Controllers;
+using Labyrinth.Console.Obstacles;
+using System.Collections.Generic;
+using System.Security.Cryptography.X509Certificates;
+
+namespace Labyrinth.Console
+{
+    public class MapGenerator
+    {
+        public MapGenerator(Dictionary<Coordinates, Obstacle> obstaclesDict, IFlowController flowController, int rightBorder, int leftBorder, int topBorder, int botBorder)
```
Comment (TonyTroeff) [inline Labyrinth.Console/MapGenerator.cs]: Merge `main` and simplify this using the incoming changes.

File: Labyrinth.Console/MapGenerator.cs [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+        private readonly int _rightBorder;
+        private readonly int _topBorder;
+        private readonly int _botBorder;
+
+        public void RenderObstacle(Obstacle obstacle)
+        {
+            this._obstaclesDict[obstacle.Coordinates] = obstacle;
+            this._flowController.RenderObstacle(obstacle);
+        }
+        public void QuickObstacleGeneration(Coordinates currentCoordinates, ObstacleEdges currentEdges)
```
Comment (TonyTroeff) [inline Labyrinth.Console/MapGenerator.cs]: I see this method for the second time now. Still, do not think it is necessary.

File: Labyrinth.Console/MapGenerator.cs [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+        {
+            Obstacle currentObstacle = new Obstacle(currentCoordinates, currentEdges);
+            this._flowController.RenderObstacle(currentObstacle);
+        }
+
+        public void GenerateXAxisBorders(int cycleStart, int cycleEnd, Coordinates topBorderLocation, Coordinates bottomBorderLocation, ObstacleEdges edgeOrientation)
```
Comment (TonyTroeff) [inline Labyrinth.Console/MapGenerator.cs]: Try to simplify this using the private fields.

**Review by TonyTroeff — CHANGES_REQUESTED** [review]

File: Labyrinth.Console/MapGenerator.cs (hunk 7 above)
Comment (ChristopherPetrov) [reply Labyrinth.Console/MapGenerator.cs]: Removed the commends.

**Review by TonyTroeff — APPROVED** [review]

**Pull request merged**